managed-pointer/cpp-20/D0567R1.md

```diff
@@ -5,222 +5,384 @@
 
-The **execution context** must encapsulate both the execution agents and a global memory region that is accessible from those execution agents. It is required that the memory region of the **execution context** be cache coherent across the execution agents created from the same invocation of an execution function. It is not required that the memory region of the **execution context** be cache coherent across execution agents created from different execution functions or on different **execution contexts**.
+Another data flow model is one that uses a runtime to create a graph of tasks, which can be optimised.
 
-*Note: some systems are capable of providing cache coherency across execution functions or **execution contexts** through the availability of shared virtual memory addressing or shared physical memory. This feature is discussed in more detail in the future work section.*
+Another data flow model is one which simply submits work to be executed and waits for it all to complete, typically called fire and forget execution.
 
-### Managed Pointer Class Template
+### 2.4. Must integrate seemlessly with exexcutors
 
-The proposed addition to the standard library is the ``managed_ptr`` class template (Figure 1). The ``managed_ptr`` is a smart pointer which has ownership of a contiguous managed allocation of memory that is shared between the host CPU and one or more **execution contexts**. It is important to note here that an **execution context** may be on the host CPU accessing the host CPU memory region, however, the interface remains the same.
+As data movement is tightly coupled with executtion, it is nessesary for a unified interface for data movement to be integrated with the unified interface for execution. This means such a unified interface must integrate seemlessly with the current design philosophy of executors, in that it must be compatible for the range of control structures which the executors abstraction supports and must support te range of platforms which executors can target.
 
-At any one time the managed memory allocation can exist in the memory regions of the host CPU and any number of **execution contexts**. However, the managed memory allocation may only be accessible in one of these memory regions at any given time. This memory region is said to be the accessible memory region. If the accessible memory region is of the host CPU, the host CPU is said to be accessible and if the accessible memory region is of an **execution context** that **execution context** is said to be the accessible **execution context**.
+At the same time a unified interface for data movement must exist at a high level of abstraction and must avoid introducing too many requirements on executor implementations, in order to avoid introducing complex requirements for heterogeneous or distributed memory architectures into C\+\+.
```

Comment:
So the intention of the `managed_ptr`, in the current proposal (not taking into consideration systems with a unified address space) is that once synchronized, it will be accessible to all execution agents of the execution context concurrently within a single execution. Currently there is no guarantees made for consistency between execution agents concurrently accessing the same address of the `managed_ptr`, i.e. the user is responsible for avoiding race conditions. The reason for this is that the exact behavior will vary from one device to another and cannot yet be fully described by the C++ memory model, so this is something that I intended to discuss in greater detail as the capabilities of devices is fleshed out in the executors proposal. However we should mention this in future work as it is very important to take into consideration.